lib/logstash/inputs/lumberjack.rb

```diff
@@ -31,47 +31,77 @@ class LogStash::Inputs::Lumberjack < LogStash::Inputs::Base
 
   # TODO(sissel): Add CA to authenticate clients with.
 
-  public
+  BUFFERED_QUEUE_SIZE = 20
+  RECONNECT_BACKOFF_SLEEP = 0.5
+  
   def register
     require "lumberjack/server"
+    require "concurrent/executors"
+    require "logstash/circuit_breaker"
+    require "logstash/sized_queue_timeout"
 
     @logger.info("Starting lumberjack input listener", :address => "#{@host}:#{@port}")
     @lumberjack = Lumberjack::Server.new(:address => @host, :port => @port,
       :ssl_certificate => @ssl_certificate, :ssl_key => @ssl_key,
       :ssl_key_passphrase => @ssl_key_passphrase)
+
+    @threadpool = Concurrent::CachedThreadPool.new(:idletime => 15)
+
+    # in 1.5 the main SizeQueue doesnt have the concept of timeout
+    # We are using a small plugin buffer to move events to the internal queue
+    @buffered_queue = LogStash::SizedQueueTimeout.new(BUFFERED_QUEUE_SIZE)
+
+    @circuit_breaker = LogStash::CircuitBreaker.new("Lumberjack input",
+                            :exceptions => [LogStash::SizedQueueTimeout::TimeoutError])
+
   end # def register
 
-  public
   def run(output_queue)
+    start_buffer_broker(output_queue)
+
     while true do
-      accept do |connection, codec|
-        invoke(connection, codec) do |_codec, line, fields|
-          _codec.decode(line) do |event|
-            decorate(event)
-            fields.each { |k,v| event[k] = v; v.force_encoding(Encoding::UTF_8) }
-            output_queue << event
+      begin
+        # Wrapping the accept call into a CircuitBreaker
+        if @circuit_breaker.close?
+          connection = @lumberjack.accept # Blocking call that creates a new connection
+
+          invoke(connection, codec.clone) do |_codec, line, fields|
+            _codec.decode(line) do |event|
+              decorate(event)
+              fields.each { |k,v| event[k] = v; v.force_encoding(Encoding::UTF_8) }
+
+              @circuit_breaker.execute { @buffered_queue << event }
+            end
           end
+        else
+          @logger.warn("Lumberjack input: the pipeline is blocked, temporary refusing new connection.")
```

Comment:
Agree, what do you think If I create a document about back pressure?
We have seen it people people will just create new issues or hack values.